File: trappingRainWater.java
import java.util.*;
public class Solution {
    public int trap(int[] A) {
        int n = A.length;
        if (n < 3)
            return 0;
        int[] leftToRight = new int[n];
        int[] rightToLeft = new int[n];
        leftToRight[0] = 0;
        rightToLeft[n-1] = 0;
        for (int i = 1; i < n; ++i) {
            leftToRight[i] = Math.max(leftToRight[i-1], A[i-1]);
            rightToLeft[n-i-1] = Math.max(rightToLeft[n-i], A[n-i]);
        }
        int total = 0;
        for (int i=1; i < n-1; i++) {
            int height = Math.min(leftToRight[i], rightToLeft[i]);
            if (height > A[i]) 
                total += height - A[i];
        }
        return total;
    }
}

File: trappingRainWater.cpp
#include "leetcode.h"
// http://blog.unieagle.net/2012/10/31/leetcode%E9%A2%98%E7%9B%AE%EF%BC%9Atrapping-rain-water/
class Solution {
public:
    int trap(int A[], int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int total = 0;
        int leftToRight[n];
        int highest = 0;
        for (int i = 0; i < n; ++i) {
            leftToRight[i] = highest;
            if (A[i] > highest) {
                highest = A[i];
            }
        }
        highest = 0;
        for (int i = n-1; i >= 0; i --) {
            if (A[i] > highest) {
                highest = A[i];
            } else if (A[i] < leftToRight[i]) {
                total += highest > leftToRight[i] ? leftToRight[i]: highest;
                total -= A[i];
            }
        }
        return total;
    }
};
int main() {
    Solution s;
    int a[] = {0,1,0,2,1,0,1,3,2,1,2,1};
    cout << s.trap(a, 12) << endl;
    int b[] = {4, 2, 0, 3, 2, 5};
    cout << s.trap(b, 6) << endl;
    return 0;
}

